ant-protocol/src/error.rs

```diff
@@ -82,3 +92,221 @@ impl From<store::Error> for Error {
+        assert_eq!(deserialized, Error::Unknown);
+    }
+
+    #[test]
+    fn test_error_retro_compatibility_reduced() {
+        // Test with a struct that has a missing variant (simulating older version)
+        #[derive(Error, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+        #[non_exhaustive]
+        enum ReducedError {
+            #[error("Chunk does not exist {0:?}")]
+            ChunkDoesNotExist(NetworkAddress),
+            #[error("Failed to deserialize hex ScratchpadAddress")]
+            ScratchpadHexDeserializeFailed,
+            // removed this variant
+            // - #[error("Failed to derive CipherText from encrypted_data")]
+            // - ScratchpadCipherTextFailed,
+            #[error("Provided cypher text is invalid")]
+            ScratchpadCipherTextInvalid,
+            #[error("There was an error getting the storecost from kademlia store")]
+            GetStoreQuoteFailed,
+            #[error("There was an error generating the payment quote")]
+            QuoteGenerationFailed,
+            #[error("Peer {holder:?} cannot find Record {key:?}")]
+            ReplicatedRecordNotFound {
+                holder: Box<NetworkAddress>,
+                key: Box<NetworkAddress>,
+            },
+            #[error("Could not Serialize/Deserialize RecordHeader to/from Record")]
+            RecordHeaderParsingFailed,
+            #[error("Could not Serialize/Deserialize Record")]
+            RecordParsingFailed,
+            #[error("The record already exists, so do not charge for it: {0:?}")]
+            RecordExists(PrettyPrintRecordKey<'static>),
+            // Missing some variants that exist in the current Error enum
+            #[error("Unknown error variant")]
+            #[serde(other)]
+            Unknown,
+        }
+
+        // Test serialization and deserialization of current Error
+        let current_error = Error::ScratchpadCipherTextInvalid;
+        let serialized_current = rmp_serde::to_vec(&current_error).unwrap();
+
+        // Test that we can deserialize into ReducedError (older version)
+        let deserialized_reduced: ReducedError =
+            rmp_serde::from_slice(&serialized_current).unwrap();
+        assert_eq!(
+            deserialized_reduced,
+            ReducedError::ScratchpadCipherTextInvalid
+        );
+
+        // Test that unknown variants fall back to Unknown
+        let unknown_variant = Error::ScratchpadCipherTextFailed;
+        let serialized_unknown = rmp_serde::to_vec(&unknown_variant).unwrap();
+        let deserialized_unknown: ReducedError =
+            rmp_serde::from_slice(&serialized_unknown).unwrap();
+        assert_eq!(deserialized_unknown, ReducedError::Unknown);
+    }
+
+    #[test]
+    fn test_error_retro_compatibility_many_missing() {
+        // test with many missing variants
+        #[derive(Error, Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
+        #[non_exhaustive]
+        enum ManyMissingVariants {
```

Comment:
to prove the compatibility best, we need :
* have this ManyMissingVariants enum contains another complex entry
* confirm such complex entry can be converted between Error and ManyMissingVariants in dual directions.